filestorage.go

```diff
@@ -149,125 +152,124 @@ func dataDir() string {
 	return filepath.Join(baseDir, "certmagic")
 }
 
-// TryLock attempts to get a lock for name, otherwise it returns
-// a Waiter value to wait until the other process is finished.
-func (fs FileStorage) TryLock(key string) (Waiter, error) {
-	fileStorageNameLocksMu.Lock()
-	defer fileStorageNameLocksMu.Unlock()
+// Lock obtains a lock named by the given key. It blocks
+// until the lock can be obtained or an error is returned.
+func (fs *FileStorage) Lock(key string) error {
+	// can't defer the unlock because we might have
+	// to Wait() for a while before returning, so we're
+	// careful to unlock at all the right places
+	fs.fileStorageNameLocksMu.Lock()
+
+	if fs.fileStorageNameLocks == nil {
+		fs.fileStorageNameLocks = make(map[string]*fileStorageWaiter)
+	}
 
 	// see if lock already exists within this process - allows
 	// for faster unlocking since we don't have to poll the disk
-	fw, ok := fileStorageNameLocks[key]
+	fw, ok := fs.fileStorageNameLocks[key]
 	if ok {
 		// lock already created within process, let caller wait on it
-		return fw, nil
+		fs.fileStorageNameLocksMu.Unlock()
+		fw.Wait()
+		return nil
 	}
 
 	// attempt to persist lock to disk by creating lock file
 
-	// parent dir must exist
 	lockDir := fs.lockDir()
-	if err := os.MkdirAll(lockDir, 0700); err != nil {
-		return nil, err
-	}
-
+	// since there isn't already a waiter for the lock, make one
 	fw = &fileStorageWaiter{
 		key:      key,
 		filename: filepath.Join(lockDir, StorageKeys.safe(key)+".lock"),
 		wg:       new(sync.WaitGroup),
 	}
+	fw.wg.Add(1)
+	fs.fileStorageNameLocks[key] = fw
+	fs.fileStorageNameLocksMu.Unlock()
 
-	var checkedStaleLock bool // sentinel value to avoid infinite goto-ing
+	for {
+		// parent dir must exist
+		if err := os.MkdirAll(lockDir, 0700); err != nil {
+			return err
+		}
 
-createLock:
-	// create the file in a special mode such that an
-	// error is returned if it already exists
-	lf, err := os.OpenFile(fw.filename, os.O_CREATE|os.O_EXCL, 0644)
-	if err != nil {
-		if os.IsExist(err) {
+		// create the file in a special mode such that an
+		// error is returned if it already exists
+		lf, err := os.OpenFile(fw.filename, os.O_CREATE|os.O_EXCL, 0644)
+		switch {
+		case err == nil:
+			// cool, we got the lock right away
+			lf.Close()
+			return nil
+		case os.IsExist(err):
 			// another process has the lock
-
-			// check to see if the lock is stale, if we haven't already
-			if !checkedStaleLock {
-				checkedStaleLock = true
-				if fs.lockFileStale(fw.filename) {
-					log.Printf("[INFO][%s] Lock for '%s' is stale; removing then retrying: %s",
-						fs, key, fw.filename)
-					os.Remove(fw.filename)
-					goto createLock
-				}
+			info, err := os.Stat(fw.filename)
+			switch {
+			case err != nil:
```

Comment:
I think this empty case will cause the loop to spin, perhaps if there is an error Stat'ing we should just assume the lock is stale, since it's apparently useless if we can't read it?